keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -231,15 +237,23 @@ type EndpointPort struct {
-- **Endpoint Port**: The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.   
+- **Endpoint Port**: The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.
 
 ### Topology (Per Endpoint)
 A new topology field (string to string map) is added to each endpoint. It can contain arbitrary topology information associated with the endpoint. If the EndpointSlice instance is derived from K8s service, the topology may contain following well known key:
 - **kubernetes.io/hostname**: the value indicates the hostname of the node where the endpoint is located. This should match the corresponding node label.
 - **topology.kubernetes.io/zone**: the value indicates the zone where the endpoint is located. This should match the corresponding node label.
 - **topology.kubernetes.io/region**: the value indicates the region where the endpoint is located. This should match the corresponding node label.
 
-If the k8s service has topological keys specified, the corresponding node labels will be copied to endpoint topology. 
+If the k8s service has topological keys specified, the corresponding node labels will be copied to endpoint topology.
+
+### Terminating (Per EndpointCondition)
+
+An optional field `Terminating` will be added to each `EndpointCondition`, indicating whether an endpoint is currently in the "Terminating" state. For the case of Pods, this would include any
+Pod that has a valid `DeletionTimestamp`. Controllers that manage EndpointSlice objects may want to opt out of this feature for optimization reasons (e.g. avoiding per-endpoint metadata expansion)
+which can be accomplished by setting the field to nil.
+
+This field allows consumers of the API to know if endpoints are currently terminating without having to initiate a watch on Pods. A specific use-case for this field is the [graceful termination handler in the IPVS proxy](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/graceful_termination.go). Because deleting an IPVS backend (a.k.a IPVS real server) immediately deletes all its corresponding conntrack entries, it can be disruptive to delete a real server during its graceful termination period. The solution today is to only delete a real server if it sees that it has no connections or it has exceeded a hardcoded duration limit (15m). This approach is not full-proof as the state of "terminating endpoints" is not preserved when kube-proxy is restarted. The `Terminating` field would remove the need for storing termintaing state if memory, improving overall design and performance of the proxy as it no longer needs to track stale backends in memory.
```

Comment:
I think we should explicitly call it in either this KEP or another the use case around graceful shutdown without disruption of pods using services for nodeport, local traffic policy, and service proxy at minimum as the guiding goal.  This actually loses some of the context - which is that we generally want to ensure that pods continue to receive existing connection traffic for all services while new connections are rapidly (or less rapidly with readiness gates) diverted away during termination.

Reply:
I'm trying to add e2e tests to kube suite that will ensure this is possible for a range of services (https://github.com/kubernetes/kubernetes/pull/89547 is one of the first changes we need to so we have a test shim)

Reply:
I'll put consumption of the `terminating` addition in a separate KEP.